ouroboros-consensus/test-storage/Test/Ouroboros/Storage/TestBlock.hs

```diff
@@ -440,10 +481,64 @@ instance UpdateLedger TestBlock where
   ledgerTipPoint = lastAppliedPoint
 
 instance HasAnnTip TestBlock where
-  -- Use defaults
+  type TipInfo TestBlock = IsEBB
+  getTipInfo = thIsEBB . unTestHeader
 
 instance ValidateEnvelope TestBlock where
-  -- Use defaults
+  -- A chain can start with an EBB with block number 0
+  firstBlockNo _ = BlockNo 0
+
+  validateEnvelope cfg oldTip hdr = do
+      when (actualBlockNo /= expectedBlockNo) $
+        throwError $ UnexpectedBlockNo expectedBlockNo actualBlockNo
+      when (actualSlotNo < expectedSlotNo) $
+        throwError $ UnexpectedSlotNo expectedSlotNo actualSlotNo
+      when (actualPrevHash /= expectedPrevHash) $
+        throwError $ UnexpectedPrevHash expectedPrevHash actualPrevHash
+      when (fromIsEBB newIsEBB && not (canBeEBB actualSlotNo)) $
+        throwError $ OtherEnvelopeError . T.pack $
+          "Unexpected EBB in slot " ++ show actualSlotNo
+    where
+      newIsEBB :: IsEBB
+      newIsEBB = thIsEBB (unTestHeader hdr)
+
+      actualSlotNo   :: SlotNo
+      actualBlockNo  :: BlockNo
+      actualPrevHash :: ChainHash TestBlock
+
+      actualSlotNo   =            blockSlot     hdr
+      actualBlockNo  =            blockNo       hdr
+      actualPrevHash = castHash $ blockPrevHash hdr
+
+      expectedSlotNo   :: SlotNo           -- Lower bound only
+      expectedBlockNo  :: BlockNo
+      expectedPrevHash :: ChainHash TestBlock
+
+      (expectedSlotNo, expectedBlockNo, expectedPrevHash) = (
+            nextSlotNo  ((annTipInfo &&& annTipSlotNo)  <$> oldTip) newIsEBB
+          , nextBlockNo ((annTipInfo &&& annTipBlockNo) <$> oldTip) newIsEBB
+          , withOrigin GenesisHash (BlockHash . annTipHash) oldTip
+          )
+
+      -- EBB shares its slot number with its successor
+      nextSlotNo :: WithOrigin (IsEBB, SlotNo) -> IsEBB -> SlotNo
+      nextSlotNo Origin          _        = SlotNo 0
+      nextSlotNo (At (IsEBB, s)) IsNotEBB = s
+      nextSlotNo (At (_    , s)) _        = succ s
+
+      -- EBB shares its block number with its predecessor. The chain can also
+      -- start with a regular block, in which case it has block number 1
+      nextBlockNo :: WithOrigin (IsEBB, BlockNo) -> IsEBB -> BlockNo
+      nextBlockNo Origin             IsEBB    = BlockNo 0
+      nextBlockNo Origin             IsNotEBB = BlockNo 1
+      nextBlockNo (At (IsNotEBB, b)) IsEBB    = b
+      nextBlockNo (At (_       , b)) _        = succ b
+
+      canBeEBB :: SlotNo -> Bool
+      canBeEBB (SlotNo s) = s `mod` epochSlots == 0
```

Comment:
Would always return `False` if EBBs are not supported (stored in the `BlockConfig`)